### PR TITLE
Ah-based SOC ledger (shadow) + anchor rework

### PR DESCRIPTION
**Pull request by danieltroger**
## What

Adds a coulomb-counting (Ah) SOC ledger running in **shadow** beside the existing Wh system, plus the anchor-detection rework it was designed around. Validated offline on 3 months of hall-sensor history before writing this code (Phase 0/0.5 backfill studies).

### Shadow (no behavior change)
- `calculated_current_2` (calibrated sensor-2 amps) published to `current_values`
- Ah ledger anchored at the latest full/empty/soft-empty event; restores ∫amps across restarts via the `raw_voltage_mv_2` subquery (full history); publishes `soc_ah` **unclamped** to `soc_values`
- Online parameter tracking, forward-only (no retroactive SOC steps): drain EMA from full→full spans (tracks the seasonal hall zero drift, 0.7 A→2.8 A spring→summer), capacity EMA from deep full↔(soft-)empty spans; persisted in new config section `soc_calculations.ah_ledger`
- NEW soft-empty anchor: falling through 49 V at |I| < 30 A ⇒ ~0.4 % SOC (validated σ ≈ 1 pp across nights)

### Production behavior changes (deliberate)
- **Full detection now uses 1-min-smoothed hall amps** instead of the inverter's `battery_current`, which under-reads (omits the inverter's own ~150 W draw) and fired the anchor minutes early
- Every anchor detection writes a `soc_anchors,type=…` marker to Influx; startup restores from markers (voltage-query fallback on first deploy) — the old voltage-only restore could adopt aborted charges as full anchors
- Planner, runtime sell/buy gates, and ws consume SOC **clamped to [0,100]** (it currently ranges −1.6…107.4 %); raw values still flow to Influx for monitoring
- Edge-triggered warn when the Wh ledgers diverge > 5 pp

## Why

Measured on 30 d of production data: SOC drifts 0.9–4.2 pp/day between anchors (worst single error +7.2 pp ≈ 4.5 kWh phantom energy for the trader). Root causes: Wh bookkeeping absorbs the charge/discharge voltage asymmetry (53.8 V in / 51.6 V out) into a wandering "parasitic" fudge, and the hall sensor's zero drifts seasonally. Backfill replay shows the Ah ledger with tracked drain reaches ~0.4–1 pp median error at anchors (~2× better than Wh under identical conditions, and immune to the voltage asymmetry by construction).

## Test plan
- [x] `yarn typecheck` (only the 2 pre-existing ads1115 x64 errors)
- [x] `backend/src/battery/ahLedgerMath.selftest.ts` (29 assertions)
- [x] `backend/src/autoTrading/planner.selftest.ts`
- [x] `planPreview.ts` dry-run (with synthetic config on the dev VM)
- [ ] Shadow period on the pi: `soc_ah` vs `average_soc` at live anchors before any cutover

Rollback: `git checkout main && sudo systemctl restart mpi-15k-controller` — the new config section is additive and harmless to leave behind.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01FShAMjWQEv4vQmqjcVqmxV

### CODE REVIEW

**Comment by danieltroger** [comment]
Deployed to the pi as the shadow test (branch checked out, service restarted 09:50 UTC in a clear window).

**First deploy immediately caught a real edge case**: the restore-time re-anchor fed parameter tracking a full→empty span whose integral was measured to *now* instead of to the anchor — 6 h of post-empty charging inflated it to an implied 949 Ah capacity. The sanity gate rejected it loudly (no state was poisoned), but a differently-timed restart could have passed the 1000–1500 gate. Fixed in 4d76093 by only tracking live-detected spans (anchor time > process start); verified on redeploy — the restore now logs `skipping parameter tracking for restored span full → empty`.

**Shadow status**: `soc_ah` publishing (25.2 % vs Wh baseline 24.5 % — 0.7 pp apart from a cold start), `calculated_current_2` flowing (119 A midday charge), anchor markers armed. First live anchors expected tonight (sell windows 17:30–20:15 UTC) / tomorrow's full charge.

File: backend/src/battery/anchorDetection.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+function publishAnchorMarker(client: AsyncMqttClient | undefined, type: AnchorType) {
+  if (!client || !client.connected) {
+    warnLog("MQTT not connected — soc_anchors marker lost; restore will fall back to the voltage queries", type);
+    return;
+  }
+  // Line protocol with a tag so Grafana can split by kind; markers are rare, so a failure is worth a log.
+  const line = `soc_anchors,type=${type} value=1`;
+  client.publish("soc_anchors", line).catch(error => warnLog("Failed to publish soc_anchors marker", type, error));
```
[the remark this answers is not in the document]
Comment (danieltroger) [reply backend/src/battery/anchorDetection.ts]: Verified, in two ways:

1. **The bridge is a wildcard subscriber.** It's telegraf on the influx VM: `[[inputs.mqtt_consumer]] topics = ["#"]` with `data_format = "influx"`, so any topic carrying line protocol is ingested and the `type` tag parses natively. (Checked before first deploy.)
2. **Empirically proven in production:** the `soc_anchors` measurement contains both live markers so far — `type=full` at 2026-07-09T15:21:38Z and `type=empty` at 2026-07-10T04:01:30Z. One honest caveat: marker-based *restore* hasn't been exercised live yet (the last restart predates the first marker, so it used the voltage fallback, which worked); the next restart is its first real test.

The literal-string drift concern is fixed in 8185195: `SOC_ANCHORS_MEASUREMENT` is now a shared constant between publish (anchorDetection) and restore (socAnchorRestore). Went with a constant rather than a config entry deliberately — the two ends must match *each other*, not be user-tunable; a config knob would add a new way to break the pairing.

File: backend/src/battery/useLastFullAndEmpty.ts [hunk 2]
```diff
@@ -1,17 +1,22 @@
...
+  smoothedBatteryCurrentAmps: Accessor<number | undefined>
+) {
   const haveSeenBatteryFullAt = createMemo<number | undefined>(prev => {
     const voltage = reactiveBatteryVoltage();
-    const current = reactiveBatteryCurrent();
-    if (voltage == undefined || current == undefined) return prev;
-    if (voltage >= config().full_battery_voltage && current < config().stop_charging_below_current) {
+    const smoothedAmps = smoothedBatteryCurrentAmps();
+    if (voltage == undefined || smoothedAmps == undefined) return prev;
+    if (fullConditionMet(voltage, smoothedAmps, config().full_battery_voltage, config().stop_charging_below_current)) {
```
[the remark this answers is not in the document]
Comment (danieltroger) [reply backend/src/battery/useLastFullAndEmpty.ts:19]: Both deliberate, and the threshold question has a clean answer: **the `< 10 A` hall-amps condition is the exact predicate the Phase 0 backfill used to construct all 75 offline anchors** (repeatability ±0.5–1 pp), so the config value was validated against hall amps by construction, not carried over blind. Directionally, the same 10 A on the hall sensor fires ~3–4 A deeper into the taper than the inverter-based check did (the inverter under-reads ~5.8% and omits its own ~3 A draw) — i.e. closer to true saturation, which was the motivation. First live full anchor (Jul 9) behaved: the Ah ledger read +1.1 pp at the anchor and the adjacent deep spans implied 1254/1261 Ah vs the offline 1240 ± 54.

On the lost inverter fallback: accepted for the shadow phase — the empty anchor (voltage-only), marker restore, and voltage-fallback restore all work without the ADC, and the energy integral shares the ADC dependency anyway. One residual edge worth a cutover-checklist item: if the ADC dies, the smoothed-amps memo holds its *last* value, so a stale <10 A reading could in principle satisfy the full condition when voltage later hits 58 V. A staleness gate (ignore smoothed amps older than a few minutes) closes it; deferring to the cutover PR to keep the mid-shadow branch minimal.

File: backend/src/battery/ahLedgerParameterTracking.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import { untrack } from "solid-js";
+import type { get_config_object } from "../config/config.ts";
+import { computeParameterUpdates, type AnchorType } from "./ahLedgerMath.ts";
+import { errorLog, logLog, warnLog } from "../utilities/logging.ts";
+
+type LedgerAnchor = { at: number; soc: number; type: AnchorType };
```
[the remark this answers is not in the document]
Comment (danieltroger) [reply backend/src/battery/ahLedgerParameterTracking.ts]: Done in 8185195 — `LedgerAnchor` now lives next to `AnchorType` in `ahLedgerMath.ts` (still a runtime-import-free module) and all three consumers import it; the `soc_anchors` string got the same treatment per the other comment.

**Comment by danieltroger** [comment]
Merging this for now to get a bit less PR stacking, LGTM after a brief read
